File: src/main/java/net/ndrei/teslacorelib/gui/ToggleButtonPiece.java
package net.ndrei.teslacorelib.gui;

import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.RenderHelper;
import net.minecraft.client.renderer.RenderItem;
import net.minecraft.item.ItemStack;
import net.ndrei.teslacorelib.inventory.BoundingRectangle;

/**
 * Created by CF on 2017-01-15.
 */
public abstract class ToggleButtonPiece extends BasicContainerGuiPiece {
    public ToggleButtonPiece(int left, int top, int width, int height) {
        super(left, top, width, height);
    }

    protected abstract int getCurrentState();
    protected abstract void renderState(BasicTeslaGuiContainer container, int state, BoundingRectangle box);
    protected abstract void clicked();

    @Override
    public void drawBackgroundLayer(BasicTeslaGuiContainer container, int guiX, int guiY, float partialTicks, int mouseX, int mouseY) {
        mouseX -= container.getGuiLeft();
        mouseY -= container.getGuiTop();
        if ((mouseX >= this.getLeft()) && (mouseY >= this.getTop()) && (mouseX <= this.getLeft() + this.getWidth()) && (mouseY <= this.getTop() + this.getHeight())) {
            container.drawFilledRect(container.getGuiLeft() + this.getLeft(), container.getGuiTop() + this.getTop(), this.getWidth(), this.getHeight(), 0x42FFFFFF);
        }
    }

    @Override
    public void drawForegroundLayer(BasicTeslaGuiContainer container, int guiX, int guiY, int mouseX, int mouseY) {
        int state = this.getCurrentState();
        int x = this.getLeft() + (this.getWidth() - 16) / 2;
        int y = this.getTop() + (this.getHeight() - 16) / 2;
        this.renderState(container, state, new BoundingRectangle(x, y, 16, 16));
    }

    protected void renderItemStack(BasicTeslaGuiContainer container, ItemStack stack, BoundingRectangle box) {
        if (stack != null) {
            RenderItem item = container.getItemRenderer();

            RenderHelper.enableGUIStandardItemLighting();
            GlStateManager.pushMatrix();
            GlStateManager.enableDepth();
            container.getItemRenderer().renderItemAndEffectIntoGUI(stack, box.getLeft(), box.getTop());
            container.getItemRenderer().renderItemOverlayIntoGUI(container.mc.fontRendererObj, stack, box.getLeft(), box.getTop(), null);
            GlStateManager.popMatrix();
            RenderHelper.disableStandardItemLighting();
            item.renderItemOverlayIntoGUI(container.getFontRenderer(), stack,
                    box.getLeft(), box.getTop(), null);
        }
    }

    @Override
    public void mouseClicked(BasicTeslaGuiContainer container, int mouseX, int mouseY, int mouseButton) {
        if (BasicContainerGuiPiece.isInside(container, this, mouseX, mouseY)) {
            this.clicked();
        }
    }
}

File: src/main/java/net/ndrei/teslacorelib/gui/ToggleButtonPiece.kt
package net.ndrei.teslacorelib.gui

import net.minecraft.client.renderer.GlStateManager
import net.minecraft.client.renderer.RenderHelper
import net.minecraft.item.ItemStack
import net.ndrei.teslacorelib.compatibility.FontRendererUtil
import net.ndrei.teslacorelib.inventory.BoundingRectangle

/**
 * Created by CF on 2017-06-28.
 */
@Suppress("unused")
abstract class ToggleButtonPiece(left: Int, top: Int, width: Int, height: Int)
    : BasicContainerGuiPiece(left, top, width, height) {

    protected abstract val currentState: Int
    protected abstract fun renderState(container: BasicTeslaGuiContainer<*>, state: Int, box: BoundingRectangle)
    protected abstract fun clicked()

    override fun drawBackgroundLayer(container: BasicTeslaGuiContainer<*>, guiX: Int, guiY: Int, partialTicks: Float, mouseX: Int, mouseY: Int) {
        var mouseX = mouseX
        var mouseY = mouseY
        mouseX -= container.guiLeft
        mouseY -= container.guiTop
        if (mouseX >= this.left && mouseY >= this.top && mouseX <= this.left + this.width && mouseY <= this.top + this.height) {
            container.drawFilledRect(container.guiLeft + this.left, container.guiTop + this.top, this.width, this.height, 0x42FFFFFF)
        }
    }

    override fun drawForegroundLayer(container: BasicTeslaGuiContainer<*>, guiX: Int, guiY: Int, mouseX: Int, mouseY: Int) {
        val state = this.currentState
        val x = this.left + (this.width - 16) / 2
        val y = this.top + (this.height - 16) / 2
        this.renderState(container, state, BoundingRectangle(x, y, 16, 16))
    }

    protected fun renderItemStack(container: BasicTeslaGuiContainer<*>, stack: ItemStack?, box: BoundingRectangle) {
        if (stack != null) {
            val item = container.itemRenderer

            RenderHelper.enableGUIStandardItemLighting()
            GlStateManager.pushMatrix()
            GlStateManager.enableDepth()
            container.itemRenderer.renderItemAndEffectIntoGUI(stack, box.left, box.top)
            container.itemRenderer.renderItemOverlayIntoGUI(FontRendererUtil.fontRenderer, stack, box.left, box.top, null)
            GlStateManager.popMatrix()
            RenderHelper.disableStandardItemLighting()
            item.renderItemOverlayIntoGUI(container.fontRenderer, stack,
                    box.left, box.top, null)
        }
    }

    override fun mouseClicked(container: BasicTeslaGuiContainer<*>, mouseX: Int, mouseY: Int, mouseButton: Int) {
        if (BasicContainerGuiPiece.isInside(container, this, mouseX, mouseY)) {
            this.clicked()
        }
    }
}
